### ISSUE TITLE
Incorrect NS3004 for derived types

### ISSUE THREAD

**Issue by samcragg** [nsubstitute/NSubstitute.Analyzers#57]
First of all, the analyzer is really useful - thanks for creating the project!

I think I'm incorrectly getting the `Could not set value of type … to argument ... because the types are incompatible` error when trying to assign a `Dictionary` to a `IReadOnlyDictionary` argument and was able to reproduce it with this code:
```C#
using NSubstitute;
using Xunit;

public class ExampleTest
{
    public interface IExampleInterface
    {
        bool Method(out object value);
    }

    [Fact]
    public void DictionaryTest()
    {
        IExampleInterface substitute = Substitute.For<IExampleInterface>();
        substitute.Method(out Arg.Any<object>())
            .Returns(ci =>
            {
                ci[0] = "string";
                return true;
            });

        bool result = substitute.Method(out object value);

        Assert.True(result);
        Assert.Equal("string", value);
    }
}
```

The line with `ci[0] = "string";` triggers the analyser error, however, the test succeeds in passing (i.e. NSubstitute does correctly set the parameter).

I think the problem is we're checking if the type is equal instead of if the assignment will work in [AbstractCallInfoAnalyzer.cs](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/632c99f9e6156ac3dd264d054aa4c039bdaff218/src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs#L248) - maybe `Equals` should be replaced with a call to [CanCast](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/632c99f9e6156ac3dd264d054aa4c039bdaff218/src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs#L66)?

**Comment by tpodolak** [comment]
@samcragg thanks for reporting. I will take a look at this once I am back from holidays. When it comes to Equals and CanCast usages I remember than Nsubstitute does different type checks depending on the context. Can you tell me which version of Nsubstitute you use?

**Comment by samcragg** [comment]
Cool, thanks for that. I've tried it in both `4.0.0-rc1` and `3.1.0` (the latest stable) against version `1.0.3` of the analyser - both times the test passes but the analyser reports an error.

**Comment by tpodolak** [comment]
Released in [1.0.4](https://www.nuget.org/packages/NSubstitute.Analyzers.CSharp/1.0.4)
